### PR TITLE
[RLlib] Cleanup SlateQ algo; add test + add target Q-net

### PR DESCRIPTION
**Pull request by sven1977**
Cleanup SlateQ algo
- add tests
- cleanup existing example script
- add target Q-net for more loss/q-value stability

## Checks

- [x] I've run `scripts/format.sh` to lint the changes in this PR.
- [ ] I've included any doc changes needed for https://docs.ray.io/en/master/.
- [ ] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [x] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested :(

### CODE REVIEW

**Comment by avnishn** [comment]
Is this branch opened off of your Move Bandits to the main folder pr #21773 . The diff is difficult to read.

**Comment by sven1977** [comment]
Yeah, it is. Hold on, let's merge the bandits branch first.
...

File: rllib/agents/slateq/slateq.py [hunk 1]
```diff
@@ -204,7 +204,9 @@ def execution_plan(workers: WorkerSet, config: TrainerConfigDict,
         # returned from the LocalReplay() iterator is passed to TrainOneStep to
         # take a SGD step.
         replay_op = Replay(local_buffer=kwargs["local_replay_buffer"]) \
-            .for_each(TrainOneStep(workers))
+            .for_each(TrainOneStep(workers)) \
+            .for_each(UpdateTargetNetwork(
```
Comment (avnishn) [inline rllib/agents/slateq/slateq.py:208]: nice!

File: rllib/env/wrappers/recsim.py [hunk 2]
```diff
@@ -175,7 +175,7 @@ def make_recsim_env(
     """
 
     class _RecSimEnv(gym.Env):
```
Comment (avnishn) [inline rllib/env/wrappers/recsim.py:177]: This should be a gym.Wrapper, not a gym.Env. It probably has no serious impacts, but I've been paranoid about this since we caught that issue with the transformed action space wrapper not seeding properly.

**Review by avnishn — COMMENTED** [review]
I have potential concerns in `rllib/env/wrappers/recsim.py` and `rllib/utils/spaces/space_utils.py`, and minor comments/questions in other places.

**Comment by bveeramani** [comment]
## :bangbang: ACTION REQUIRED :bangbang:

We've switched our code formatter from YAPF to Black (see #21311).

To prevent issues with merging your code, here's what you'll need to do:

1. Install Black
```
pip install -I black==21.12b0
```

2. Format changed files with Black
```
curl -o format-changed.sh https://gist.githubusercontent.com/bveeramani/42ef0e9e387b755a8a735b084af976f2/raw/7631276790765d555c423b8db2b679fd957b984a/format-changed.sh
chmod +x ./format-changed.sh
./format-changed.sh
rm format-changed.sh
```

3. Commit your changes.
```
git add --all
git commit -m "Format Python code with Black"
```

4. Merge master into your branch.
```
git pull upstream master
```

5. Resolve merge conflicts (if necessary).

After running these steps, you'll have the updated `format.sh`.

File: rllib/agents/slateq/slateq_torch_model.py [hunk 3]
```diff
@@ -0,0 +1,211 @@
...
+from ray.rllib.utils.typing import ModelConfigDict, TensorType
+
+torch, nn = try_import_torch()
+F = None
+if nn:
+    F = nn.functional
+
+
+class QValueModel(nn.Module):
+    """The Q-value model for SlateQ"""
```
[the remark this answers is not in the document]
Comment (sven1977) [reply rllib/agents/slateq/slateq_torch_model.py:16]: I linked the paper in the docstring.

File: rllib/agents/slateq/slateq_torch_model.py [hunk 4]
```diff
@@ -0,0 +1,211 @@
...
+        if user.shape[-1] == 0:
+            x = doc_flat
+        else:
+            user_repeated = user.repeat(num_docs, 1)
+            x = torch.cat([user_repeated, doc_flat], dim=1)
+        x = self.layers(x)
+        # Similar to Google's SlateQ implementation in RecSim, we force the
+        # Q-values to zeros if there are no clicks.
+        x_no_click = torch.zeros((batch_size, 1), device=x.device)
+        return torch.cat([x.view((batch_size, num_docs)), x_no_click], dim=1)
```
[the remark this answers is not in the document]
Comment (sven1977) [reply rllib/agents/slateq/slateq_torch_model.py:76]: I think this is described in the paper. We always compute the individual Q-values for each of the (possible) documents + 1 no-click placeholder. In case the user doesn't click on any document, we need to keep track of the value of that "empty doc", I guess.

File: rllib/env/wrappers/recsim.py (hunk 2 above)
Comment (sven1977) [reply rllib/env/wrappers/recsim.py:180]: Great catch! Will change.

Comment (sven1977) [reply rllib/env/wrappers/recsim.py:180]: Done, saved a bunch of boilerplate :).

File: rllib/agents/slateq/slateq_torch_model.py [hunk 5]
```diff
@@ -0,0 +1,211 @@
...
+        batch_size, num_docs, embedding_size = doc.shape
+        doc_flat = doc.view((batch_size * num_docs, embedding_size))
+        # No user features.
+        if user.shape[-1] == 0:
+            x = doc_flat
+        else:
+            user_repeated = user.repeat(num_docs, 1)
+            x = torch.cat([user_repeated, doc_flat], dim=1)
+        x = self.layers(x)
+        # Similar to Google's SlateQ implementation in RecSim, we force the
```
[the remark this answers is not in the document]
Comment (sven1977) [reply rllib/agents/slateq/slateq_torch_model.py:71]: Done.

**Review by gjoliver — COMMENTED** [review]
I am kinda curious if this has ever worked since it's introduced? so many updates.

File: rllib/agents/slateq/slateq.py [hunk 6]
```diff
@@ -64,11 +64,15 @@
...
+        # Must be SlateEpsilonGreedy to deal with the fact that e.g.:
+        # action_space=MultiDiscrete([5, 5]) <- slate-size=2, num-docs=5
+        # but action distribution is Categorical(5*4 / 2)
+        # -> all possible unique slates.
+        "type": "SlateEpsilonGreedy",
         # Config for the Exploration class' constructor:
         "initial_epsilon": 1.0,
         "final_epsilon": 0.02,
-        "epsilon_timesteps": 10000,  # Timesteps over which to anneal epsilon.
+        "epsilon_timesteps": 50000,  # Timesteps over which to anneal epsilon.
```
[the remark this answers is not in the document]
Comment (sven1977) [reply rllib/agents/slateq/slateq.py:75]: I'm actually not sure our existing SlateQ would really work w/o all of these enhancements made in this PR.
But yeah, fair enough, for safety reasons, we can leave it at 10k and do 50k in the tuned_examples.

File: rllib/utils/spaces/space_utils.py [hunk 7]
```diff
@@ -307,9 +307,12 @@ def convert_element_to_space_type(element: Any, sampled_element: Any) -> Any:
     def map_(elem, s):
         if isinstance(s, np.ndarray):
             if not isinstance(elem, np.ndarray):
-                raise ValueError(
-                    "Element should be of type np.ndarray but is instead of \
-                        type {}".format(type(elem)))
+                if s.shape == ():
```
[the remark this answers is not in the document]
Comment (sven1977) [reply rllib/utils/spaces/space_utils.py:323]: Hold on `s` here is in fact a numpy array (see if clause one line above the other one):
```
        if isinstance(s, np.ndarray):  # <- `s` is numpy, e.g. np.array(5.5).
            if not isinstance(elem, np.ndarray):  # <- but `elem` isn't, e.g. float(5.5).
                if s.shape == ():
```

[the remark this answers is not in the document]
Comment (sven1977) [reply rllib/utils/spaces/space_utils.py:323]: Yeah, `elem` should NOT be a torch tensor. Happy to add a check for this. ...

[the remark this answers is not in the document]
Comment (sven1977) [reply rllib/utils/spaces/space_utils.py:323]: Normally, we call this function with `elem` being a sample from a `gym.spaces.Space`, so numpy. But in the RecSim env case, this may be a python float. That's why I added this additional option here.

[the remark this answers is not in the document]
Comment (sven1977) [reply rllib/utils/spaces/space_utils.py:323]: done

File: rllib/agents/slateq/tests/test_slateq.py [hunk 8]
```diff
@@ -0,0 +1,42 @@
...
+
+import ray
+import ray.rllib.agents.slateq as slateq
+from ray.rllib.examples.env.recsim_recommender_system_envs import \
+    LongTermSatisfactionRecSimEnv
+from ray.rllib.utils.test_utils import check_compute_single_action, \
+    check_train_results, framework_iterator
+
+
+class TestSlateQ(unittest.TestCase):
```
[the remark this answers is not in the document]
Comment (sven1977) [reply rllib/agents/slateq/tests/test_slateq.py:15]: Hey, sorry, but this would take another day or so to write due to the 2 different models and complex loss math used. 

Could we defer such a test for later? Note that this algo was originally contributed to Ray by Henry (@desktable ) in late 2020 and this PR only adds a few enhancements to it to make the Q-loss and -value terms more stable and to enable exploration. It will require some more investigation by us in the coming weeks to fully claim support for this and confirm "hard-task" learning capabilities.

[the remark this answers is not in the document]
Comment (sven1977) [reply rllib/agents/slateq/tests/test_slateq.py:15]: But I think it's great that you want to go through the paper to double check with our implementation! Please let me know, if you find more issues with our math.

**Review by gjoliver — APPROVED** [review]
nice pr and cleanups.
